### PR TITLE
[SPARK-42876][SQL] DataType's physicalDataType should be private[sql]

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

`physicalDataType` should not be a public API but be private[sql].

### Why are the changes needed?

This is to limit API scope to not expose unnecessary API to be public.

### Does this PR introduce _any_ user-facing change?

No since we have not released Spark 3.4.0 yet.

### How was this patch tested?

N/A

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala [hunk 1]
```diff
@@ -119,7 +119,7 @@ abstract class DataType extends AbstractDataType {
 
   override private[sql] def acceptsType(other: DataType): Boolean = sameType(other)
 
-  def physicalDataType: PhysicalDataType = UninitializedPhysicalType
+  private[sql] def physicalDataType: PhysicalDataType = UninitializedPhysicalType
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala:122]: can we change the modifier in subclasses as well?

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DataType.scala:122]: done

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.4!

**Comment by cloud-fan** [comment]
also cc @xinrong-meng this is from api auditing

**Pull request closed without merging**